### PR TITLE
Fix spatial loss map step labeling

### LINKED ISSUE #747: [Bug] Spatial loss map titles and saved PDFs mislabeled in on_test_epoch_end when val_steps_to_log contains out-of-range steps
### Describe the Bug

In `ForecasterModule.test_step` (`neural_lam/models/module.py:657-664`), `log_spatial_losses` filters `val_steps_to_log` to include only steps within the evaluation rollout horizon (`step <= spatial_loss.shape[1]`):

```python
# neural_lam/models/module.py:657-664
log_spatial_losses = spatial_loss[
    :,
    [
        step - 1
        for step in val_steps_to_log
        if step <= spatial_loss.shape[1]
    ],
]
self.spatial_loss_maps.append(log_spatial_losses)
```

However, during test epoch aggregation in `on_test_epoch_end` (`neural_lam/models/module.py:1014-1049`), the aggregated tensor `mean_spatial_loss` (which only contains the filtered subset of steps) is zipped against the **raw, unfiltered** `val_steps_to_log`:

```python
# neural_lam/models/module.py:1014-1025
loss_map_figs = [
    vis.plot_spatial_error(
        error=loss_map,
        datastore=self.datastore,
        title=f"Test loss, t={t_i} "
        f"({(self.time_step_int * t_i)} {self.time_step_unit})",
    )
    for t_i, loss_map in zip(
        val_steps_to_log,
        mean_spatial_loss,
    )
]
```

---

### Impact & Failure Modes

When `--val_steps_to_log` contains any step that exceeds the rollout length (for example, `--val_steps_to_log 100 1 5` with `--ar_steps_eval 25`):

1. **Mislabeled Figures and Saved Artifacts**:
   - In `test_step`, only steps `1` and `5` are extracted (yielding 2 spatial slices along the time dimension).
   - In `on_test_epoch_end`, `zip([100, 1, 5], [slice_step_1, slice_step_5])` pairs:
     - Step 1 spatial loss with `t_i = 100`. The figure is titled `Test loss, t=100` and written to `spatial_loss_maps/loss_t100.pdf`.
     - Step 5 spatial loss with `t_i = 1`. The figure is titled `Test loss, t=1` and written to `spatial_loss_maps/loss_t1.pdf`.
     - Step 5 is completely dropped from the output iteration.

2. **Silent Scientific Misattribution**:
   - The logged heatmaps and exported PDF loss maps display wrong forecast horizons without crashing or alerting the user.

---

### Minimal Reproducible Example

```python
import torch

val_steps_to_log = [100, 1, 5]
spatial_loss = torch.randn(2, 25, 100)  # (B, pred_steps=25, N)

# test_step filtering logic
logged_steps = [
    step - 1 for step in val_steps_to_log if step <= spatial_loss.shape[1]
]
log_spatial_losses = spatial_loss[:, logged_steps]  # shape (2, 2, 100) -> contains step 1 and step 5

mean_spatial_loss = torch.mean(log_spatial_losses, dim=0)  # shape (2, 100)

# on_test_epoch_end pairing logic
for t_i, _ in zip(val_steps_to_log, mean_spatial_loss):
    print(f"Assigned title t={t_i}")
```

**Output:**
```text
Assigned title t=100  # Actually step 1
Assigned title t=1    # Actually step 5
```

---

### Proposed Fix

Filter `val_steps_to_log` in `on_test_epoch_end` matching the condition in `test_step` before plotting and saving:

```python
valid_val_steps = [
    step for step in val_steps_to_log if step <= spatial_loss_tensor.shape[1]
]
```
and zip against `valid_val_steps` when generating `loss_map_figs` and `pdf_loss_map_figs`.

### PR DESCRIPTION
**Pull request by vikramsingh920**
## Describe your changes
Fixes an issue where out-of-range values in `val_steps_to_log` could cause
spatial loss maps to be associated with the wrong forecast steps.

< Summary of the changes.>

Updated `test_step` to keep track of the filtered evaluation steps and use them
when generating spatial loss map titles and PDF filenames. Added a regression
test to verify that out-of-range steps are skipped and valid steps are labeled
correctly.

< Please also include relevant motivation and context. >

When `val_steps_to_log` contains steps outside the test rollout length, those
steps are filtered out when calculating the spatial loss. However, the original
`val_steps_to_log` was still used when labeling and saving the loss maps. This
could cause the loss maps to be associated with the wrong forecast steps and
PDF filenames.

< List any dependencies that are required for this change. >
No new dependencies are required.
## Issue Link
fixes https://github.com/mllam/neural-lam/issues/747
< Link to the relevant issue or task, if applicable > (e.g. `closes #00` or `solves #00`)

## Type of change

- [✅] 🐛 Bug fix (non-breaking change that fixes an issue)
- [ ] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📖 Documentation (Addition or improvements to documentation)

## Checklist before requesting a review

- [ ] My branch is up-to-date with the target branch - if not update your fork with the changes from the target branch (use `pull` with `--rebase` option if possible).
- [✅] I have performed a self-review of my code
- [✅] For any new/modified functions/classes I have added docstrings that clearly describe its purpose, expected inputs and returned values
- [ ] I have placed in-line comments to clarify the intent of any hard-to-understand passages of my code
- [ ] I have updated the [README](README.MD) to cover introduced code changes
- [✅] I have added tests that prove my fix is effective or that my feature works
- [✅] I have given the PR a name that clearly describes the change, written in imperative form ([context](https://www.gitkraken.com/learn/git/best-practices/git-commit-message#using-imperative-verb-form)).
- [ ] I have requested a reviewer and an assignee (assignee is responsible for merging). This applies only if you have write access to the repo, otherwise feel free to tag a maintainer to add a reviewer and assignee.

## Checklist for reviewers

Each PR comes with its own improvements and flaws. The reviewer should check the following:
- [ ] the code is readable
- [ ] the code is well tested
- [ ] the code is documented (including return types and parameters)
- [ ] the code is easy to maintain

## Author checklist after completed review

- [✅] I have added a line to the CHANGELOG describing this change, in a section
  reflecting type of change (add section where missing):
  - *added*: when you have added new functionality
  - *changed*: when default behaviour of the code has been changed
  - *fixes*: when your contribution fixes a bug
  - *maintenance*: when your contribution is relates to repo maintenance, e.g. CI/CD or documentation

## Checklist for assignee

- [ ] PR is up to date with the base branch
- [ ] the tests pass
- [ ] (if the PR is not just maintenance/bugfix) the PR is assigned to the next milestone. If it is not, propose it for a future milestone.
- [ ] author has added an entry to the changelog (and designated the change as *added*, *changed*, *fixed* or *maintenance*)
- Once the PR is ready to be merged, squash commits and merge the PR.

### CODE REVIEW

**Comment by GiGiKoneti** [comment]
Thanks @vikramsingh920. The core synchronization via `_test_spatial_steps` and the regression test look good, but a few items need fixing:

1. **Revert unrelated files:** Please revert `neural_lam/config.py` and `neural_lam/train_model.py` (`add_safe_globals` is out of scope for #747).
2. **Fix `ty` type checking error in `neural_lam/models/module.py` (lines 890–893):** Line 891 `self.logger.save_dir` fails `ty` because the ignore comment was placed on line 892 (`exist_ok=True,`). If directory creation was only needed for the test fixture, create the directory in the test instead of modifying `create_metric_log_dict`.
3. **Guard lower bound in `test_step`:** Use `1 <= step <= spatial_loss.shape[1]` instead of `step <= spatial_loss.shape[1]` to prevent non-positive steps (`0`, `-1`) from silently slicing from the end via negative indexing in programmatic setups.
4. **Rebase after #746:** Per @sadamov's note in #747, please rebase on `main` once #746 lands.
5. **Clean up diff:** Please revert the tuple unpacking change in `tests/test_plotting.py:558`.

cc : @sadamov

**Comment by vikramsingh920** [comment]
Thanks for the review @GiGiKoneti. I have addressed the points in the latest push:

1. Reverted `neural_lam/config.py` and `neural_lam/train_model.py`.
2. Fixed the `ty` error by creating the directory inside the test fixture rather than `create_metric_log_dict`.
3. Added the `1 <= step <= spatial_loss.shape[1]` lower-bound guard in `test_step` and expanded regression test coverage to include non-positive steps (`-1, 0`).
4. Reverted the tuple unpacking syntax in `tests/test_plotting.py:558`.
5. Will rebase on `main` once #746 lands.

**Comment by GiGiKoneti** [comment]
two small issues remain:

1. **`black` and `flake8` failures:** `pre-commit run flake8` fails on lines exceeding 80 chars in `neural_lam/models/module.py:659` and `tests/test_plotting.py:928, 955`, and `black` reformats `module.py`.
2. **Tuple unpacking at `tests/test_plotting.py:558`:** Line 558 still drops the parentheses around `(init_states, ...)`; please revert to keep the diff clean.

**Comment by vikramsingh920** [comment]
Thanks @GiGiKoneti, I addressed both comments in the latest push:

1. Wrapped long lines in neural_lam/models/module.py and tests/test_plotting.py so flake8 and black pass cleanly.
2. Restored the tuple unpacking parentheses at tests/test_plotting.py:558 to keep the diff clean.

**Comment by vikramsingh920** [comment]
Hi @GiGiKoneti  and @sadamov , when you get a chance, could you please take a look at my PR https://github.com/mllam/neural-lam/pull/752 and let me know if there’s anything I should change or improve? I’d really appreciate your feedback. Thanks!

**Comment by GiGiKoneti** [comment]
Since #746 is still open, this PR will need a rebase on `main` once #746 lands before it can be merged. Leaving final review and sequencing to Simon.

**Comment by vikramsingh920** [comment]
hii @GiGiKoneti and @sadamov , #746 has now been merged, so I’ve rebased #752 onto the latest `main` and addressed the previous review comments.
Could you please take a look at #752 when you get a chance? Thanks!
